arxiv/integration/api/service.py

```diff
@@ -0,0 +1,261 @@
+    re-used.
+
+    To implement an HTTP API service integration, define a class that inherits
+    from :class:`.HTTPIntegration`. At minimum, it should include an inner
+    class called ``Meta`` with the attribute ``service_name``; this is used to
+    identify the configuration parameters for the service. For example:
+
+    .. code-block:: python
+
+       class MyCoolIntegration(HTTPIntegration):
+           class Meta:
+               service_name = "cool_beans"
+
+           def get_something(self, anything: str, token: str) -> dict:
+               data, _, _ = self.json("get", f"/thing/{anything}", token)
+               return data
+
+
+    and in your ``config.py``, include:
+
+    .. code-block:: python
+
+       COOL_BEANS_ENDPOINT = "http://thecool.beans.arxiv.org:9876"
+       COOL_BEANS_VERIFY = False   # (or True, if you are using SSL)
+
+
+   Thanks to :class:`arxiv.integration.meta.MetaIntegration`, you should be
+   able to use this integration in your app like:
+
+   .. code-block:: python
+
+      from my.cool.services import MyCoolIntegration
+      from arxiv.users import auth
+
+      app = Flask(__name__)
+      MyCoolIntegration.init_app(app)
+      auth.Auth(app)
+
+      @app.route('/foo')
+      def foo():
+          MyCoolIntegration.get_something('foo', request.environ['token'])
+          ...
+
+
+    """
+
+    class Meta:
+        """
+        Additional info about this class.
+
+        This should be overridden by any child class.
+        """
+
+        service_name = "base"
+
+    def __init__(self, endpoint: str, verify: bool = True,
+                 headers: dict = {}) -> None:
+        """
+        Initialize an HTTP session.
+
+        Parameters
+        ----------
+        endpoint : str
+            Base service URI.
+        verify : bool
+            Whether or not SSL certificate verification should enforced.
+        headers : dict
+            Headers to be included on all requests.
+
+        """
+        self._session = requests.Session()
+        self._verify = verify
+        self._retry = Retry(  # type: ignore
+            total=10,
+            read=10,
+            connect=10,
+            status=10,
+            backoff_factor=0.5
+        )
+        self._adapter = requests.adapters.HTTPAdapter(max_retries=self._retry)
+        self._session.mount(f'{urlparse(endpoint).scheme}://', self._adapter)
+        if not endpoint.endswith('/'):
+            endpoint += '/'
+        self._endpoint = endpoint
+        self._session.headers.update(headers)
+
+    def _path(self, path: str, query: dict = {}) -> str:
+        """Generate a full path for a request."""
+        o = urlparse(self._endpoint)
+        path = path.lstrip('/')
+        return urlunparse((  # type: ignore
+            o.scheme, o.netloc, f"{o.path}{path}",
+            None, urlencode(query), None
+        ))
+
+    def _check_status(self, resp: requests.Response,
```

Comment:
The HTTP status code handling could be incorporated in a more composable way. It seems that it could be a function such as http_status_to_raised_exception(status:int)->None. Then it could be used anywhere.

Reply:
Good idea; pulled this out into its own function